### PR TITLE
5 Upfront and maintance costs

### LINKED ISSUE #5: Upfront costs function

### LINKED ISSUE #6: Maintenance costs function

### LINKED ISSUE #15: Getter for gas boiler installation costs

### PR DESCRIPTION
**Pull request by sofiapinto**
---

# Description

This PR:
- adds a getter for gas boiler installation costs
- creates a script with functions for computing maintenance and upfront costs for a given technology, as a total and per year of operation.
- adds a notebook for experimenting the functions above

Closes #15 #5 #6 

# Instructions for Reviewer

There are still a few things to discuss in our next stand up, but here's a first version of this work for review. Please check 
- if the logic of the code makes sense
- if there's anything you would change
- if the functions work for you in the notebook

# Checklist:

- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    if (type(subsidy_model_or_value) is str) and (heating_system == "ashp"):
+        subsidy_value = get_ashp_subsidy_value(
+            subsidy_model=subsidy_model_or_value, installation_year=installation_year
+        )
+        return installation_cost - subsidy_value
+    else:  # float provided
+        return installation_cost - subsidy_model_or_value
+
+
+def compute_total_lifetime_costs(
```
Comment (sofiapinto) [inline asf_lifetime_cost_model/pipeline/cost_computations.py:211]: this could obviously just take 3 parameters: maintenance, running and installation costs and sum them. But for now i'll leave it like this, as I don't know how we're organising these functions.

File: asf_lifetime_cost_model/getters/data_getters.py [hunk 2]
```diff
@@ -410,3 +410,16 @@ def get_ashp_subsidy_options_data() -> pd.DataFrame:
         bucket_name="asf-lifetime-cost-model",
         s3_key="inputs/ashp_subsidy_options.csv",
     )
+
+def get_gas_boiler_installation_costs() -> pd.DataFrame:
```
Comment (eglucas) [inline asf_lifetime_cost_model/getters/data_getters.py:414]: Thanks for adding the csv to the S3 bucket and writing this additional data getter!

The gas boiler installation cost csv table has different archetype names in the column `archetype_label` to the names in the table for ASHP installation costs (not your fault as these tables have inconsistent labelling in the spreadsheet model). 

We could either change the gas boiler installation cost table to match the archetype names in the ASHP installation cost table, or we could have a separate discussion to modify the archetype names because even the ones in the ASHP table aren't super clear.

I know that our preference is to refer to the archetypes by their full label name, rather than their archetype number - but for consistency with the ASHP installation cost table, I think we should also add the `archetype_number` column to the gas boiler table. Just in case there is a need to refer to the archetypes by their number further down the line.

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    """Computes the total maintenance cost for technology given the annual maintenance cost and the expected lifetime.
+
+    Args:
+        annual_maintenance_costs (float): The annual maintenance cost of the heating system.
+        life_span (int): Number of years the heating system is assumed to be operational.
+
+    Returns:
+        float: The total maintenance costs over the specified number of years.
+    """
+    return annual_maintenance_costs * life_span
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]:
```suggestion
def compute_total_maintenance_cost(maintenance_cost: float, maintenance_frequency_per_year: float, life_span: int) -> float:
    """Computes the total maintenance cost for technology given the cost of each maintenance session, number of times the heating system is serviced each year and the expected lifetime.

    Args:
        maintenance_cost (float): The cost of one maintenance servicing session for heating system.
        maintenance_frequency_per_year (float): Average number of times the heating system is serviced each year. 
        life_span (int): Number of years the heating system is assumed to be operational.

    Returns:
        float: The total maintenance costs over the specified number of years.
    """
    return maintenance_cost * maintenance_frequency_per_year * life_span
```

For clarity and transparency, I suggest that we separate out `annual_maintenance_costs` to have separate inputs for cost of 1 maintenance session (`maintenance_cost`) and number of maintenance sessions in each year (`maintenance_frequency_per_year`). Feel free to edit these variable names if there are clearer alternatives!

I think it would be good to have this flexibility to track separately how the mission's assumptions evolve on the cost of maintenance and the number of times heating systems are serviced. For example, if the mission wants to model that heat pumps are serviced every other year, then the frequency can be 0.5 but it will be clear that the cost of 1 maintenance session is the same.

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+
+def creates_cost_reduction_data(
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]:
```suggestion
def create_cost_reduction_data(
```

Not sure if using 'creates' instead of 'create' was an intentional naming convention style - but think it's clearer to stay consistent across all function names :)

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    Raises:
+        ValueError: If the heating system inputed is not supported or the year is above or equal to 2035.
+
+    Returns:
+        float: Upfront cost of installing a heating system.
+    """
+    # Get installation cost for a specific heating system, archetype, and decile (where applicable)
+    installation_cost = get_installation_cost(heating_system=heating_system, archetype=archetype, decile=decile)
+
+    cost_reduction_data = creates_cost_reduction_data(annual_cost_reduction=annual_cost_reduction)
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]:
```suggestion
    cost_reduction_data = create_cost_reduction_data(annual_cost_reduction=annual_cost_reduction)
```

As above

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    Returns:
+        dict: A dictionary with years as keys and cost values as values.
+    """
+    cost_per_year = cost_value / life_span
+    cost_breakdown = dict.fromkeys(
+        range(installation_year, installation_year + life_span),
+        cost_per_year,
+    )
+
+    return cost_breakdown
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]:
```suggestion
def create_annualised_cost_time_series(cost_value: float, life_span: int, installation_year: int) -> dict:
    """Creates a time series of annualised cost per year in the lifetime of the heating system.

    Args:
        cost_value (float): The total cost value to be annualised.
        life_span (int): Number of years the heating system is assumed to be operational.
        installation_year (int): The year in which the heating system is installed.

    Returns:
        dict: A dictionary with years as keys and cost values as values.
    """
    cost_per_year = cost_value / life_span
    annualised_cost_time_series = dict.fromkeys(
        range(installation_year, installation_year + life_span),
        cost_per_year,
    )

    return annualised_cost_time_series
```

There's a distinction between "cost breakdown in each year of lifetime" and the "annualised lifetime cost".

The `cost_per_year` calculation you've coded here where the lifetime `cost_value` is divided by number of years (`life_space`) is the annualised cost (i.e. total lifetime cost averaged over the number of years in lifetime). I've made some rewording suggestions to make this distinction clearer.

The annualised cost calculation is different to the actual cost of owning a heat pump/gas boiler in each year of its lifetime. In this alternative time series, the total lifetime cost will be broken down by which year it is actually spent (like a yearly schedule of costs) - for example:
- upfront installation cost will only be attributed to the first year of its lifetime
- running costs will be different each year because of the DESNZ wholesale price projections in each future year

I don't think creating this other time series cost breakdown function is necessary for us to do, unless the policy team want this "actual total cost in each year of the heating system lifetime" as an output. This is something worth checking with the policy team in case they want this communicated as a figure or downloadable table.

File: asf_lifetime_cost_model/pipeline/cost_computations.py (hunk 1 above)
Comment (eglucas) [reply asf_lifetime_cost_model/pipeline/cost_computations.py:211]: I like this way you've coded it! I think it makes sense for the arguments of this function to be all the individual arguments into the underlying upfront/maintenance/running cost functions. I like that it will mean that all inputs are listed in one place,

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    subsidy_value = subsidy_data[subsidy_data["model"] == subsidy_model][str(installation_year)].iloc[0]
+
+    return subsidy_value
+
+
+def compute_upfront_cost(
+    heating_system: str,
+    archetype: str,
+    annual_cost_reduction: float,
+    installation_year: int,
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]: In my running costs code, I've called the first year of the heating system's lifetime the `purchase_year`. I'm happy with either `installation_year` or `purchase_year` - do you have any strong preferences?

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    decile: int = None,
+    subsidy_model_or_value: Union[str, float] = 0,
+) -> float:
+    """Computes the upfront costs of installing a heating system.
+
+    Args:
+        heating_system (str): heating system.
+            Takes "ashp" (for air source heat pump) or "boiler" (for gas boiler).
+        archetype (str): Property archetype
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py:142]: I agree that it is more readable to refer to a property archetype by its label name. I will amend the running costs code to take the archetype label as an input instead of the archetype number.

File: asf_lifetime_cost_model/analysis/notebooks/cost_computations_demo.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+# The cost of installing an air source heat pump, WITHOUT subtracting the subsidy from the upfront cost:
+
+# %%
+cost_computations.compute_upfront_cost(
+    heating_system="ashp",
+    archetype="pre_1950_semi_terraced_house_5_rooms",
+    annual_cost_reduction=0.01,
+    installation_year=2026,
+    decile=50,
+    subsidy_model_or_value="flat",
```
Comment (eglucas) [inline asf_lifetime_cost_model/analysis/notebooks/cost_computations_demo.py]:
```suggestion
    subsidy_model_or_value=0,
```

I think you forgot to adjust the arguments here to demonstrate the case without the subsidy. I'm assuming setting the `subsidy_model_or_value` to 0 is the only way to exclude the subsidy?

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+        heating_system=heating_system,
+        archetype=archetype,
+        decile=decile,
+        annual_cost_reduction=annual_cost_reduction,
+        installation_year=installation_year,
+        subsidy_model_or_value=subsidy_model_or_value,
+    )
+    maintenance_cost = compute_total_maintenance_cost(
+        annual_maintenance_costs=annual_maintenance_costs, life_span=life_span
+    )
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]: If you agree on my suggestion above to replace `annual_maintenance_costs` with two separate arguments for `maintenance_cost` and `maintenance_frequency_per_year` then these argument changes will have to be reflected here

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    Args:
+        heating_system (str): heating system.
+            Takes "ashp" (for air source heat pump) or "boiler" (for gas boiler).
+        archetype (str): Property archetype
+            Takes: TODO: add final list of archetypes.
+        decile (int, optional): cost decile, only used for air source heat pumps.
+        annual_cost_reduction (float): how much the cost of installing the heating system reduces each year
+            in comparison to the previous year.
+        installation_year (int): the year in which the heating system is installed.
+        subsidy_model_or_value (Union[str, float], optional): The subsidy model to get the subsidy to be subtracted
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]: I think this is a really neat way of handling different possibilities for the subsidy scenarios!

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+            OR a fixed subsidy value.
+            Models include: "flat", "slow stepdown", "fast stepdown", "high", "zero from 2028",
+            "smallest", "no subsidy".
+
+    Raises:
+        ValueError: If the heating system inputed is not supported or the year is above or equal to 2035.
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py:160]: Not sure if you were intending to explicitly write the ValueError messages for these cases, but they're currently missing.

When I input:
- a year later than 2035 in `installation_year` (e.g. 2036), it returns `KeyError: 2036`
- `subsidy_model_or_value="flat"` with `heating_system="boiler"`, it returns `UFuncTypeError: ufunc 'subtract' did not contain a loop with signature matching types (dtype('float64'), dtype('<U4')) -> None`

I think it would be good to explicitly code the ValueError messages for these two different cases, in case it's not clear for future users why the inputs are being rejected.

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+            "smallest", "no subsidy".
+
+    Raises:
+        ValueError: If the subsidy model is not applicable for the heating system.
+
+    Returns:
+        float: The total costs over the specified number of years.
+    """
+    if (heating_system != "ashp") and (subsidy_model_or_value in config.get("ashp_subsidy_options")):
+        raise ValueError("Subsidy model is only applicable for air source heat pumps. Please provide a number.")
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]: Currently, the ValueError is only triggered if the `subsidy_model_or_value` is in the list of valid options.

I think it could be useful (to really safeguard against potential user errors in the future) to also have a ValueError handling the case where `heating_system != "ashp"` and an _unrecognised_ `subsidy_model_or_value` is provided. 

At present, if I input `heating_system=boiler` and a typo for `subsidy_model_or_value="fla"` then the error message returned is: `UFuncTypeError: ufunc 'subtract' did not contain a loop with signature matching types (dtype('float64'), dtype('<U3')) -> None`. In this case, it would be useful to have a specific error message explaining that a number is needed because the heating system is not an air source heat pump.

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+    Raises:
+        ValueError: If the subsidy model is not supported.
+
+    Returns:
+        float: The subsidy value for the specified model.
+    """
+    supported_models = config.get("ashp_subsidy_options")
+
+    if subsidy_model not in supported_models:
+        raise ValueError(f"Unsupported subsidy model: {subsidy_model}. Supported models are: {supported_models}")
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py:118]: Just flagging that the ValueError message does list "other" as a supported subsidy model because it's in config, but if you set `subsidy_model="other` then inevitably you get an index error because it's not present in the table.

Not sure if it is worth integrating the possibility of entering the custom subsidy values into this function now (as I know we want that functionality in the streamlit app), or wait a bit until we know more about how we want to format the custom subsidy values in the streamlit code?

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+        raise ValueError("Decile must be a multiple of 10 between 10 and 90, inclusive.")
+
+    if heating_system == "ashp":
+        costs_data = data_getters.get_ashp_installation_costs()
+        cost_value = costs_data[costs_data["archetype_label"] == archetype][f"cost_percentile_{decile}"].iloc[0]
+    elif heating_system == "boiler":
+        costs_data = data_getters.get_gas_boiler_installation_costs()
+        cost_value = costs_data[(costs_data["archetype_label"] == archetype)]["cost"].iloc[0]
+    else:
+        raise ValueError(f"Unsupported heating system: {heating_system}")
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]:
```suggestion
        raise ValueError(f"Unsupported heating system: {heating_system}. Supported heating systems are {list(config.get("life_span_default").keys())}"
```

Suggest to add a line on which heating system strings are valid. I'm assuming that any new heating system we introduce later will also have its own lifespan that will be added to the `life_span_default` definition in config.

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+    decile: int = None,
+    subsidy_model_or_value: Union[str, float] = 0,
+) -> float:
+    """Computes the upfront costs of installing a heating system.
+
+    Args:
+        heating_system (str): heating system.
+            Takes "ashp" (for air source heat pump) or "boiler" (for gas boiler).
+        archetype (str): Property archetype
+            Takes: TODO: add final list of archetypes.
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py:143]: Just making a note that the final list of archetype names will have to be the ones that are standardised across both the gas boiler and ashp installation cost tables

**Review by eglucas — COMMENTED** [review]
Thank you so much for this great work Sofia!! All the code and functions are really easy to follow and are structured in a logical way. Was a pleasure to review.

For this PR, I ran all the code in the demo notebook and all the functions worked as expected. I also used this notebook to test edge cases for invalid arguments. I also went through all the code in `cost_computations.py`, and inspected the output of the new `get_gas_boiler_installation_costs()` getter.

In my comments below, I've made some suggestions on changes to improve the flexibility of the maintenance cost function, modify the naming and wording in the total cost breakdown function and add some more ValueError messages to handle particular edge cases. I've also flagged some differences in how things are coded here vs how I've coded them in the running costs function, as well as the difference in archetype labels between the gas boiler and ashp installation cost tables.

Let me know what you think of the suggestions and I'm happy to look at any other resulting changes. Once this is all merged, I think that integrating the running costs computation code into `cost_computations.py` as well as the `compute_total_lifetime_costs()` function will be fairly straightforward!

File: asf_lifetime_cost_model/getters/data_getters.py (hunk 2 above)
Comment (sofiapinto) [reply asf_lifetime_cost_model/getters/data_getters.py:414]: Yep - I opened this issue for that specifically: https://github.com/nestauk/asf_lifetime_cost_model/issues/19 

Agree with **modifying the names to make them clearer** and with **adding an archetype number column**.

File: asf_lifetime_cost_model/pipeline/cost_computations.py (hunk 3 above)
Comment (sofiapinto) [reply asf_lifetime_cost_model/pipeline/cost_computations.py]: Nice, I like that. I think `maintenance_frequency_per_year` should be an int though, rather than a float. And we could default it to 1 i.e. once per year.

File: asf_lifetime_cost_model/pipeline/cost_computations.py (hunk 4 above)
Comment (sofiapinto) [reply asf_lifetime_cost_model/pipeline/cost_computations.py]: Yep - it's a typo, thank you! :)

File: asf_lifetime_cost_model/pipeline/cost_computations.py (hunk 6 above)
Comment (sofiapinto) [reply asf_lifetime_cost_model/pipeline/cost_computations.py]: Thank you for the clarification, so important!

File: asf_lifetime_cost_model/pipeline/cost_computations.py (hunk 7 above)
Comment (sofiapinto) [reply asf_lifetime_cost_model/pipeline/cost_computations.py]: I like `purchase_year` :)

File: asf_lifetime_cost_model/analysis/notebooks/cost_computations_demo.py (hunk 9 above)
Comment (sofiapinto) [reply asf_lifetime_cost_model/analysis/notebooks/cost_computations_demo.py]: For no subsidy you can use "no subsidy". I actually need to change this to take a dictionary of subsidy values per year rather than just one value.

File: asf_lifetime_cost_model/pipeline/cost_computations.py (hunk 11 above)
Comment (sofiapinto) [reply asf_lifetime_cost_model/pipeline/cost_computations.py]: I need to make it `Union[str, dict]` so that rather than one value, a value per year can be inputed.

File: asf_lifetime_cost_model/pipeline/cost_computations.py (hunk 12 above)
Comment (sofiapinto) [reply asf_lifetime_cost_model/pipeline/cost_computations.py:160]: absolutely! thanks, will add.

File: asf_lifetime_cost_model/pipeline/cost_computations.py (hunk 13 above)
Comment (sofiapinto) [reply asf_lifetime_cost_model/pipeline/cost_computations.py]: agree - will change!

File: asf_lifetime_cost_model/pipeline/cost_computations.py (hunk 14 above)
Comment (sofiapinto) [reply asf_lifetime_cost_model/pipeline/cost_computations.py:118]: Yes, good shout, I need to change this. I didn't think ahead, now that I've started coding the app I see that this doesn't work.

File: asf_lifetime_cost_model/pipeline/cost_computations.py (hunk 3 above)
Comment (eglucas) [reply asf_lifetime_cost_model/pipeline/cost_computations.py]: Great! Yes, I agree about the default being 1 - but I still think there may be benefit to allowing it to be a float since we can have a frequency that is e.g. every other year (0.5) or every four years (0.25). It may be the case later that we find out that heat pumps are often serviced less than once a year.

Comment (sofiapinto) [reply asf_lifetime_cost_model/pipeline/cost_computations.py]: ahh, fair enough. I'll leave it like this then.

**Comment by sofiapinto** [comment]
Hey @eglucas, thanks so much for the thorough review and for the explanations about the domain. Really appreciate those.

Changes include:
- Adding `maintenance_frequency_per_year` to the calculation of total maintenance costs;
- Renaming `create_cost_breakdown_per_year_in_lifetime() `to `create_annualised_cost_time_series()` to match what is being calculated
- Renamed `installation_year` to `purchase_year`
- Change `subsidy_model_or_value` so that it takes a model or a dictionary of values. Can you please check the logic still makes sense?
- Added ValueError messages where they were missing and improved existing ones
- Changed the demo notebook to match the changes in the code + added some new tests (to check the errors)

I haven't changed the archetype names or added the ID, as I opened a separate issue to deal with that. In the issue I also highlighted where changes are required. Feel free to add if I missed anything! https://github.com/nestauk/asf_lifetime_cost_model/issues/19

Could you please take a final look to check that you're happy about the changes I made and the the notebook works as expected for you.

Thanks in advance :)

P.s.: As I mentioned, I anticipate a few changes when I create the class. But i'll merge this first.

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+
+
+def get_ashp_subsidy_value(subsidy_model: str, purchase_year: int) -> float:
+    """Gets the subsidy value for a specific subsidy model.
+
+    Args:
+        subsidy_model (str): The model of the subsidy.
+            Models include: "flat", "slow stepdown", "fast stepdown", "high", "zero from 2028",
+            "smallest", "no subsidy".
+        purchase_year (int): The year in which the heating system is installed.
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]:
```suggestion
        purchase_year (int): The year in which the heating system is purchased and installed.
```

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+    return upfront_cost + maintenance_cost
+
+
+def create_annualised_cost_time_series(cost_value: float, life_span: int, purchase_year: int) -> dict:
+    """Creates a time series of annualised cost per year in the lifetime of the heating system.
+
+    Args:
+        cost_value (float): The total cost value to be annualised.
+        life_span (int): Number of years the heating system is assumed to be operational.
+        purchase_year (int): The year in which the heating system is installed.
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]:
```suggestion
        purchase_year (int): The year in which the heating system is purchased and installed.
```

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+    cost_reduction_data = create_cost_reduction_data(annual_cost_reduction=annual_cost_reduction)
+    cost_reduction_value = cost_reduction_data[purchase_year]
+
+    installation_cost = installation_cost * cost_reduction_value
+
+    if (type(subsidy_model_or_input_values) is str) and (heating_system == "ashp"):
+        subsidy_value = get_ashp_subsidy_value(subsidy_model=subsidy_model_or_input_values, purchase_year=purchase_year)
+        return installation_cost - subsidy_value
+    elif heating_system == "ashp":  # float provided
+        return installation_cost - subsidy_model_or_input_values[purchase_year]
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]:
```suggestion
    elif heating_system == "ashp":  # dictionary provided
        if purchase_year not in subsidy_model_or_input_values:
            raise ValueError(f"Please provide subsidy value for purchase year: {purchase_year}")
        return installation_cost - subsidy_model_or_input_values[purchase_year]
```

Suggest adding (yet another haha) ValueError message if a subsidy value isn't provided for the purchase year.

The logic for when a dictionary is provided makes sense to me though!

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+
+    if (
+        (heating_system == "ashp")
+        and type(subsidy_model_or_input_values) is str
+        and subsidy_model_or_input_values not in config.get("ashp_subsidy_options")
+    ):
+        supported_models = config.get("ashp_subsidy_options")
+        raise ValueError(
+            f"Unsupported subsidy model: {subsidy_model_or_input_values}. Supported models are: {supported_models}."
+        )
```
Comment (eglucas) [inline asf_lifetime_cost_model/pipeline/cost_computations.py]: I think that both of these ValueErrors should be in the `compute_upfront_cost()` function.

The first error where heating_system = boiler and a valid subsidy model is given: this is currently missing from `compute_upfront_cost()`.

The second one where heating_system = ashp and an invalid subsidy model is given:  this is already in `compute_upfront_cost()` via using `get_ashp_subsidy_value()`.

My suggestion is to move the first error handle (boiler + valid subsidy model given) to `compute_upfront_cost()`, and remove both error handling blocks from `compute_total_lifetime_cost()`. I think this might be tidier because all errors related to invalid arguments will then be handled within each cost component function.

**Review by eglucas — COMMENTED** [review]
Thanks for making the changes, Sofia! Looks good!

I've added a few last suggestions on rewording documentation and (sorry!) a couple of final ValueError tweaks. 

None of them are crucial though, so if you don't agree/think they are necessary then happy for you to merge!

**Comment by eglucas** [comment]
Hi @sofiapinto ✨ I have added code that allows the option for purchasing with a loan.

The modified functions are:
- `compute_upfront_cost`: This now takes the additional arguments `life_span`, `purchase_with_loan`, `loan_interest_rate`. Please can you check the logic of how the different combinations of subsidy/loan/interest rate are handled. FYI: The formula for `annual_loan_payment` is the same as what is in the spreadsheet model where the [PMT Excel function](https://support.microsoft.com/en-gb/office/pmt-function-0214da64-9a63-4996-bc20-214433fa6441) is used.
- `compute_total_lifetime_cost`: Arguments are updated to now also include the new arguments required for `compute_upfront_cost`. 

I have also added some more demo and testing examples for purchasing with a loan in `cost_computations_demo.py`. Please could you run those and check they're working as expected. Please also test any other cases that may break the upfront costs function!

File: asf_lifetime_cost_model/analysis/notebooks/cost_computations_demo.py [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+    subsidy_model_or_input_values="flat",
+    purchase_with_loan=False,
+)
+
+interest = upfront_cost_with_loan - upfront_cost_without_loan
+interest
+
+# %%
+# Annualised
+1909 / 15
```
Comment (eglucas) [inline asf_lifetime_cost_model/analysis/notebooks/cost_computations_demo.py]: FYI - this £127 annualised cost of interest is comparable to what I have calculated for the affordability progress indicator. In the progress indicator calculations, purchasing an ASHP with 5% interest rate over 15 year repayment period (current BUS subsidy, average ASHP installation cost, medium Typical Domestic Consumption Values for energy use) - our estimate is £133 annualised loan interest.

File: asf_lifetime_cost_model/pipeline/cost_computations.py [hunk 22]
```diff
@@ -191,6 +192,8 @@ def compute_upfront_cost(
         upfront_cost = installation_cost
 
     if purchase_with_loan:
+        if life_span is None:
+            raise ValueError("Please provide `life_span` in years when purchasing with loans.")
```
Comment (sofiapinto) [inline asf_lifetime_cost_model/pipeline/cost_computations.py:199]: Made life_span optional and added this ValueError. Let me know if you disagree @eglucas

**Review by sofiapinto — COMMENTED** [review]
Hey @eglucas - thank you so much for adding the code with the option for purchasing with a loan. Looks great!

I implemented your suggestions and only made one change which is highlighted in the comment below.

Let me know what you think. thanks in advance!
